packages/cozy-konnector-libs/libs/signin.js

```diff
@@ -0,0 +1,146 @@
+/**
+ * The goal of this function is to provide an handy method to log the user in,
+ * on html form pages. On success, it resolves a promise with a parsed body.
+ *
+ * Errors:
+ *
+ * - LOGIN_FAILED if the validate predicate is false
+ * - INVALID_FORM if the element matched by `formSelector` is not a form or has
+ *   no `action` attribute
+ * - UNKNOWN_PARSING_STRATEGY if `parseStrategy` is not one of the accepted
+ *   values : `raw`, `cheerio`, `json`.
+ * - VENDOR_DOWN if a request throws a RequestError
+ *
+ * It does not submit values provided through `select` tags, except if populated
+ * by user with `formData`.
+ * Limit between `baseUrl` and `page` is important. The form `action` will be
+ * concatenated the same way `page` is, to constitute the submission url. For
+ * now, a form with an absolute `action` is not comptabile with this function.
+ *
+ * - `pageUrl` is the url to access the html form
+ *
+ * - `formSelector` is used by cheerio to uniquely identify the form in which to
+ *   log in
+ *
+ * - `formData` is an object of { name: value, … }. It is used to populate the
+ *   form, in the proper inputs with the same name as the properties of this
+ *   object, before submitting it.
+ *
+ * - `parseStrategy` allow the user to resolve `signin` with a preparsed body.
+ *   The choice of the strategy for the parsing is one of : `raw`, `json` or
+ *   `cheerio`. `cheerio` being the default.
+ *
+ * - `validate` is a predicate takin two arguments `statusCode` and
+ *   `parsedBody`. If it is false, `LOGIN_FAILED` is thrown, otherwise the
+ *   konnector continues.
+ *
+ * - `opts` allows to pass eventual options to the `signin`'s `requestFactory`.
+ *   It could be useful for pages using `latin1` `encoding` for instance.
+ *
+ * @module signin
+ */
+const errors = require('../helpers/errors')
+const rerrors = require('request-promise/errors')
+const log = require('cozy-logger').namespace('cozy-konnector-libs')
+const requestFactory = require('./request')
+const url = require('url')
+
+module.exports = function signin (
+  pageUrl,
+  formSelector,
+  formData,
+  parse = 'cheerio',
+  validate = defaultValidate,
+  opts = {})
```

Comment:
`requestOpts` would be more explicit

And in the best case I would get : 

```javascript
signin('https://http://quotes.toscrape.com/login', 'form', {username: 'test', password: 'passw0rd'})
```

Still, I prefer named parameters with an option object.

Handling default values and mandatory parameters would be less straightforward but I think the connector developers would find it easier to use.

Reply:
> requestOpts would be more explicit

92b36d5

> Still, I prefer named parameters with an option object.

So are you ok with 026d748 ?

Reply:
Why not , from connector developer point of view : 

```javascript
signin(
  url: 'https://http://quotes.toscrape.com/login',
  formSelector: 'form',
  formData: {
    username: 'test',
    password: 'passw0rd'
})
```

?

Reply:
If I am not wrong, he would have to wrap everything in an object like so:

```js
signin({
  url: 'https://http://quotes.toscrape.com/login',
  formSelector: 'form',
  formData: {
    username: 'test',
    password: 'passw0rd'
  }
})
```

If you confirm this is what you want, I will commit tomorrow ;-)

Reply:
0b8009a